crates/biome_js_analyze/src/semantic_analyzers/correctness/use_hook_at_top_level.rs

```diff
@@ -100,14 +121,159 @@ fn enclosing_function_if_call_is_at_top_level(call: &JsCallExpression) -> Option
         .and_then(|body| body.parent::<AnyJsFunction>())
 }
 
+/// Model for tracking which function calls are preceeded by an early return.
+#[derive(Clone, Default)]
+struct EarlyReturnsModel(HashMap<JsCallExpression, TextRange>);
+
+impl Deref for EarlyReturnsModel {
+    type Target = HashMap<JsCallExpression, TextRange>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for EarlyReturnsModel {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[derive(Default)]
+struct EarlyReturnDetectionVisitor {
+    early_returns: EarlyReturnsModel,
+    stack: Vec<EarlyReturnDetectionVisitorStackEntry>,
+}
+
+#[derive(Default)]
+struct EarlyReturnDetectionVisitorStackEntry {
+    early_return: Option<TextRange>,
+}
```

Comment:
nit: 

```rs
struct EarlyReturnDetectionVisitorStackEntry(Option<TextRange>);
```

Reply:
I wanted to be a bit explicit about what the text range applies to, because we care about text ranges for different purposes here. I think the previous conversation already showed some confusion on front :sweat_smile:

Reply:
I mean, the name of the struct already has "EarlyReturn", I think it's already explicit enough 🤔